### ISSUE TITLE
Which configuration did you use for IMU data?

### ISSUE THREAD

**Issue by jjzazuet** [IvanVnucec/UKF_STM32F4#4]
Hi. This is a cool project!

I'm also trying to track noisy input (single 3d vectors) using a Kalman filter.

In your test, it looks like you're feeding IMU data to the filter using the same configuration used to validate the functionality of `ukfLib`:

https://github.com/IvanVnucec/UKF_STM32F4/blob/9c65e89ebc16464c812d043e82c2c8b0f3090d66/src/main.c#L136

Shouldn't that source line be using the second filter configuration?

Specifically:

https://github.com/IvanVnucec/UKF_STM32F4/blob/main/src/ukfLib/cfg/ukfCfg2.c

I'm trying to find out what set of Fx and Hy functions to implement to track a single 3 coordinate vector's inputs.

Thanks!

**Comment by IvanVnucec** [comment]
Hey, I'm glad you like it. 
Its been a while since I've worked on the project so I'm not sure about the `UkfMatrixCfg` or `UkfMatrixCfg2` you've mentioning. Looking at the code, especially comments headers in the `src/ukfLib/cfg/ukfCfg.c` and the `src/ukfLib/cfg/ukfCfg2.c` files, you can see that they definetly differ  in the State vector representation. In the `ukfCfg.c` file, we have quaternion as a vector embedded in the State vector, while in `ukfCfg2.c` file we have four quaternion elements and estimated bias in the gyro measurements. For more info about each state representation, look into links provided in the comments header of both ukf cfg files.

As for the `void ukf_test2(void)` function, there, I obviously am not testing the second type of UKF and I am not sure why. 
Could you test out the second UKF cfg and see what results you're getting. I'm not sure how you would test the estimated results...

**Comment by jjzazuet** [comment]
Thanks for the notes @IvanVnucec !

In my case, I made a Java port of ivo-georgiev's UKF library, here: https://github.com/vaccovecrana/jukf/tree/feature/micro

I also ported the reference test case for the textbook example he cites in his wiki pages:

- Wiki page: https://github.com/ivo-georgiev/ukfLib/wiki/Tracking-example(Simon,-2006)-:-CFG0
- Test (passing): https://github.com/vaccovecrana/jukf/blob/feature/micro/src/test/java/io/vacco/jukf/UkfTestConfig.java

In my use case, an object detection library I'm working on [here](https://github.com/vaccovecrana/kimaris/) gives me:

- Approximate `X/Y` pixel coordinates where an object is detected.
- The diameter `S` of a circle where most of the detected object is contained.

My problem is that these detection outputs (a vector with X/Y/S values) are jittery.

In this example, I'm detecting the face and eyes of a person.

https://user-images.githubusercontent.com/1002433/227106835-ac7dc767-144a-4853-8900-e1e41ec5ef95.mp4

So I'd like to explore the possibility of smoothing out the detection signal using a Kalman filter.

What I don't know is what kind of problem equations I can use to model this problem as `FxNN` and `HyNN` functions that Ivo's library can use to process the signal (i.e. matrix sizes and functions for the Ukf configuration object).

Any help or references to literature are appreciated!

Otherwise, if the question doesn't apply to this project, let me know and I'll close the issue.

Thanks again!

**Comment by IvanVnucec** [comment]
@jjzazuet 

Interesting problem. While I've understood your problem, I can't give you the answer to your problem regarding finding `FxNN` and `Hynn`. Obviously your outputs `Yk` are x, y of the point and circle coordinates, and I'm not so sure about your inputs `Xk+1`. From there you could derive UKF.
Kalman filters are quite hard topics and they involve that you know the theory behind it in order to derive equations used to solve a problem. I remember that I would copy-paste a solution from internet and it would not work. After number of tries, I've went studying Kalman filters. If you're interested, I've liked [this](https://github.com/rlabbe/Kalman-and-Bayesian-Filters-in-Python) source/book very much.

As for the specifics of your problem, I'm not sure if the Kalman filter is the best solution to you. Did you done research about tracking features between the frames and then filtering those?

**Comment by jjzazuet** [comment]
Hi @IvanVnucec , thanks so much for the book reference! I'll check it out.

Now, on the matter of tracking intra-frame features, yes I did consider at some point applying a variant of the [Lucas-Kanade stabilization method](https://cseweb.ucsd.edu//classes/sp02/cse252/lucaskanade81.pdf) not on pixel values, but rather on the raw `x/y/s` coordinate outputs of the object detection stage.

At some point I remember seeing that it also depended on quite a few matrix operations (similar to the Kalman Filter), so I decided to explore the latter method first. I think I should re-consider exploring the former for my problem.

With that said, I think this issue can be closed.

Thanks again and cheers!
